lib/Twig/Extension/Core.php

```diff
@@ -821,10 +821,32 @@ function _twig_default_filter($value, $default = '')
 function twig_get_array_keys_filter($array)
 {
     if ($array instanceof Traversable) {
-        return array_keys(iterator_to_array($array));
+        if ($array instanceof IteratorAggregate) {
+            $array = $array->getIterator();
+        } // else $array must be an instance of IteratorAggregate here, we could call getIterator on that one etc. until we get an iterator
+
+        if ($array instanceof Iterator) {
+            $keys = [];
+            $array->rewind();
+            while($array->valid()) {
+                $keys[] = $array->key();
+                $array->next();
+            }
+
+            return $keys;
+        }
+
+        $keys = [];
+        foreach ($array as $key => $item) {
+            $keys[] = $key;
+        }
+
+        return $keys;
     }
 
     if (!is_array($array)) {
+        @trigger_error('Using "keys" must be done on an array or Traversable, deprecated since version 1.29 and will be removed in Twig 2.0', E_USER_DEPRECATED);
```

Comment:
So something like:
`$twigParams['c'] = function(){ return 1;};`

`{% for a,b in c %}`

is legit?